src/coreclr/src/jit/lsraxarch.cpp

```diff
@@ -1364,7 +1367,18 @@ int LinearScan::BuildBlockStore(GenTreeBlk* blkNode)
             // The srcAddr must be in a register.  If it was under a GT_IND, we need to subsume all of its
             // sources.
             dstAddrRegMask = RBM_RDI;
-            srcRegMask     = RBM_RSI;
+            if (src->isContained())
+            {
+                srcRegMask = RBM_RSI;
+            }
+            else
+            {
+                assert(compiler->compNoReturnRetyping());
+                assert(src->OperIs(GT_BITCAST) && src->AsUnOp()->gtGetOp1()->IsCall());
+                // It could only happen if we the source was a struct call,
+                // we don't have a source address, we just need to generate one move.
+                srcRegMask = RBM_LNGRET;
+            }
```

Comment:
I think these changes add excess complexity and confusion to this method. I'd suggest to extract the case where the rhs is a call returning values in registers into a separate method.

Reply:
Done, thanks.